bn_mp_pack.c

```diff
@@ -1,49 +1,50 @@
 #include "tommath_private.h"
-#ifdef BN_MP_EXPORT_C
+#ifdef BN_MP_PACK_C
 /* LibTomMath, multiple-precision integer library -- Tom St Denis */
 /* SPDX-License-Identifier: Unlicense */
 
 /* based on gmp's mpz_export.
  * see http://gmplib.org/manual/Integer-Import-and-Export.html
  */
-mp_err mp_export(void *rop, size_t *countp, int order, size_t size,
-                 int endian, size_t nails, const mp_int *op)
+mp_err mp_pack(void *rop, size_t maxcount, size_t *writtencount, mp_order order, size_t size,
+               mp_endian endian, size_t nails, const mp_int *op)
 {
    mp_err err;
-   size_t odd_nails, nail_bytes, i, j, bits, count;
+   size_t odd_nails, nail_bytes, i, j, count, written = 0;
    unsigned char odd_nail_mask;
 
    mp_int t;
 
+   if (maxcount == 0u) {
+      return MP_BUF;
+   }
+
    if ((err = mp_init_copy(&t, op)) != MP_OKAY) {
       return err;
    }
 
-   if (endian == 0) {
-      union {
-         unsigned int i;
-         char c[4];
-      } lint;
-      lint.i = 0x01020304;
-
-      endian = (lint.c[0] == '\x04') ? -1 : 1;
+   if (endian == MP_NATIVE_ENDIAN) {
+      MP_GET_ENDIANNESS(endian);
    }
 
    odd_nails = (nails % 8u);
    odd_nail_mask = 0xff;
-   for (i = 0; i < odd_nails; ++i) {
+   for (i = 0u; i < odd_nails; ++i) {
       odd_nail_mask ^= (unsigned char)(1u << (7u - i));
    }
    nail_bytes = nails / 8u;
 
-   bits = (size_t)mp_count_bits(&t);
-   count = (bits / ((size * 8u) - nails)) + (((bits % ((size * 8u) - nails)) != 0u) ? 1u : 0u);
+   count = mp_pack_count(&t, nails, size);
 
-   for (i = 0; i < count; ++i) {
-      for (j = 0; j < size; ++j) {
+   for (i = 0u; i < count; ++i) {
+      if (i >= maxcount) {
```

Comment:
I just asked myself if it wouldn't make more sense to do this check before allocating the copy? who's interested in partial results anyway?

Reply:
Sure, but this is how we defined it for the other functions returning MP_BUF too. We return how many bytes/units have been written. The issue is that there is no quick check if the result will fit for to_radix. So we could avoid writing partial results for ubin/sbin/pack. For to_radix we necessarily have to allow partial results. And this is a bit inconsistent. We could also say that in the case MP_BUF is returned, the buffer is in some kind of inconsistent state and the result should not be used. Then we can do whatever.

Reply:
> We could also say that in the case MP_BUF is returned, the buffer is in some kind of inconsistent state and the result should not be used

I think we should do that if it makes our code easier! ... and to_radix can stay an exception with well-defined output ... @czurnieden your thoughts?

Reply:
Well this could even apply to to_radix. The difference is only that to_radix does not allow for an immediate bail out.

Reply:
AFAIU it won't make the code of to_radix easier, so we simply leave it as it is.